src/e3/testsuite/driver/diff.py

```diff
@@ -0,0 +1,314 @@
+import os
+import re
+import sys
+
+from e3.diff import diff
+from e3.os.fs import unixpath
+from e3.testsuite.result import Log, binary_repr
+from e3.testsuite.driver.classic import ClassicTestDriver, TestAbortWithError
+
+
+def indent(text, prefix="  "):
+    """prepend ``prefix`` to every line in ``text``.
+
+    :param str text: text to transform.
+    :param str prefix: string to prepend.
+    :rtype: str
+    """
+    return "\n".join((prefix + line) for line in text.split("\n"))
+
+
+class OutputRefiner(object):
+    """
+    Interface to refine a test output before baseline and actual comparison.
+
+    Sometimes, the way a library/tool works forces it to have outputs that
+    depends on the environment (for instance: the location of the testsuite on
+    the filesystem, the current date, etc.). Refiners makes it possible for
+    a testsuite to "hide" these discrepancies during the diff computation.
+
+    Note that output refiners might get called bytes strings when the test
+    drivers operate in binary mode.
+    """
+
+    def refine(self, output):
+        """
+        Refine a test/baseline output.
+
+        :param str output: Output to refine.
+        :rtype: str
+        """
+        raise NotImplementedError
+
+
+class RefiningChain(OutputRefiner):
+    """Simple wrapper for a sequence of output refiners applied in chain."""
+
+    def __init__(self, refiners):
+        self.refiners = refiners
+
+    def refine(self, output):
+        for r in self.refiners:
+            output = r.refine(output)
+        return output
+
+
+class Substitute(OutputRefiner):
+    """Replace substrings in outputs."""
+
+    def __init__(self, substring, replacement=""):
+        """
+        :param str substring: Substring to replace.
+        :param str replacement: Replacement to use for the substitution.
+        """
+        self.substring = substring
+        self.replacement = replacement
+
+    def refine(self, output):
+        return output.replace(self.substring, self.replacement)
+
+
+class PatternSubstitute(OutputRefiner):
+    """Replace patterns in outputs."""
+
+    def __init__(self, pattern, replacement=""):
+        """
+        :param str pattern: Pattern (regular expression) to replace.
+        :param str replacement: Replacement to use for the substitution.
+        """
+        self.regexp = re.compile(pattern)
+        self.replacement = replacement
+
+    def refine(self, output):
+        return self.regexp.sub(self.replacement, output)
+
+
+def ReplacePath(path, replacement=""):
```

Comment:
Don't use camel case for functions

Reply:
The fact that it’s a function is just a detail for users, so I would like to keep the naming scheme consistent across output refiners. I’ll turn this into a class, instead.